pystac/collection.py

```diff
@@ -675,13 +675,37 @@ def from_dict(
 
         return collection
 
-    def get_assets(self) -> Dict[str, Asset]:
-        """Get this item's assets.
+    def get_assets(
+        self,
+        media_type: Optional[Union[str, pystac.MediaType]] = None,
+        role: Optional[Union[str, pystac.MediaType]] = None,
+    ) -> Dict[str, Asset]:
+        """Get this collection's assets.
+
+        Args:
+            media_type: If set, filter the assets such that only those with a
+                matching ``media_type`` are returned.
+            role: If set, filter the assets such that only those with a matching
+                ``role`` are returned.
 
         Returns:
-            Dict[str, Asset]: A copy of the dictionary of this item's assets.
+            Dict[str, Asset]: A dictionary of assets that match ``media_type``
+                and/or ``role`` if set or else all of this collection's assets.
         """
-        return dict(self.assets.items())
+        assets = dict(self.assets.items())
+        if media_type is not None:
+            assets = {
+                key: asset
+                for key, asset in assets.items()
+                if asset.media_type == media_type
+            }
+        if role is not None:
+            assets = {
+                key: asset
+                for key, asset in assets.items()
+                if asset.roles is not None and role in asset.roles
+            }
+        return assets
```

Comment:
This is a little nitpicky, but we're iterating over the assets as many as three times; once to create the new dict, then twice if both `media_type` and `role` are set. Iterating once seems preferable (with a short circuit):

```python
if media_type is None and role is None:
    return dict(self.assets.items())
assets = dict()
for key, asset in self.assets.items():
    if (media_type is None or asset.media_type == media_type) and (role is None or role in asset.roles):
        assets[key] = asset
return assets
```

Reply:
Nope, not nitpicky. That's cleaner. We do need to augment the condition to guard against a `None` value in `asset.roles`:

```python
if (media_type is None or asset.media_type == media_type) and (role is None or (asset.roles is not None and role in asset.roles)):
```

That's getting to be a mouthful. We could simplify `asset.roles is not None and ...` to just `asset.roles and ...`, but that's not as clear (to me).